### DISCUSSION TITLE
Random bugs, drop outs & poor performance --> Solved

### DISCUSSION THREAD

**Discussion by natbesh** [syssi/esphome-jk-bms#235] [General]
Finally all up & running - but boy it was a journey.
Its a superb repo & works incredibly well - as long as you understand the limitations of the systems involved.

Me: Used ESP's for ages first foray into ESPHome - so off to a bit of a rocky start.

I couldnt get RS485 working, so went to BLE - nothing but issues.
Random dropouts of BLE & WiFi, really poor performance, kept getting dropouts when flashing firmware - just a horrible time & nothing consistent - so next to impossible to debug & work through.

The solution - seems that its all down to the performance capabilities of the hardware.
I followed the steps & used the "example" config in its entirety on a WROOM 32 - it worked a bit, then would randomly drop out for 5min & back again. Kept giving WiFi errors - right next to a good station & still issues.

It seems  that the hardware (ESP32 & possibly the BMS) is just not capable of doing updates for "all" sensors in the config every 5 seconds, pull them from BLE & then send them over Wifi.

If you're having random issues, dropouts, performance issues my advice:
 - Adjust the polling frequency to 10-15seconds, just make the ESP32 & BMS do less work
 - Cut the config down to basically nothing, in terms of sensors & slowly add them over time:
   - Necessary cell voltages
   - Power, current & total voltages
   - other switches & widgets
 
Hope that helps some others out - i'll put a pull request in to update the examples with more info/descriptions.

**Comment by chfink0** [comment]
Thank you for your suggestions, wery appreciated. Could you please specify how you adjust the polling frequency? Is it the throttle setting?

**Reply by syssi** [reply]
Yes. It's the `throttle` setting!

**Reply by chfink0** [reply]
Excellent, thank you!

**Comment by lycaki** [comment]
Nice catch, mine does the same, just trying ... be great if it fixes as it's annoying!

**Comment by dr3amr** [comment]
I use an ESP32 CAM (this is what I had in "stock") and with throttle at 2s it works perfect. I use bellow cards and I can see all information updating at 2s. Possibly it depends also on the ESP32 chip type.

https://github.com/syssi/esphome-jk-bms/discussions/230

**Reply by austin202220** [reply]
interesting... I ordered a couple now to test with. I dont know much about what the specs mean but seems like it has a more powerful processor than a plain old es32-wrover.

**Comment by natbesh** [comment]
Just a follow-up note on this...  
I got mine working well polling 16 cells & a few other parameters from the BMS.  I added a NEEY balancer - and polling the same number of metrics from 2 BLE devices & its randomly dropping out & has poor performance.

As per here: https://esp32.com/viewtopic.php?t=6707
The BLE & WiFi share the same radio - antenna, circuits etc, so there's going to be some contention.

@syssi - not sure of the control possible in the code, but is it possible to sleep wifi, poll BLE, sleep BLE & send WiFi data?
Looks like i'll be putting in more ESPs!

**Reply by syssi** [reply]
I recommend avoiding WiFi if you are using Bluetooth. There a some ESP32 boards equipped with a Ethernet port (https://esphome.io/components/ethernet.html#configuration-examples). This will resolve all concurrency issues.

**Comment by magnusboman** [comment]
@syssi When using a board with ethernet, will it default to that when eth config is present or is there a config option for it to use ethernet?

**Reply by syssi** [reply]
Just remove the `wifi` section and add `ethernet` section/configuration

**Comment by EasilyBoredEngineer** [comment]
please see https://github.com/syssi/esphome-jk-bms/discussions/364 for some fairly effective workarounds I have devised to address these issues with BLE.

**Reply by natbesh** [reply]
SUPERB!!!
The reset on BLE is a bit hacky, but hey - gotta do what you need. I've just added multiple ESP's to avoid the crash happening. It seems to be ok up to about 20 or so sensors.

The turn off BLE on OTA is golden - nothing hacky about that!  nice work!

**Comment by kev300** [comment]
I have no issues with bms ble throttle: 0s and soyosource component + a lot of other stuff in my config lika garage door control, ads and ina measurements, etc.
The ESP32 also gets new power consumption values each second via wifi.

Only the OTA update was buggy, so I added an automation, which disables BLE as soon as an OTA update starts.

https://github.com/kev300/HA_with_EspHome/blob/main/esp-home_config/garage.yaml
